dlrover/python/master/shard_manager/batch_dataset_manager.py

```diff
@@ -187,3 +186,169 @@ def restore_checkpoint(self, checkpoint: DatasetShardCheckpoint):
+        logger.info(
+            "Assign task %s of dataset %s to worker %s",
+            task.task_id,
+            self._dataset_splitter.dataset_name,
+            worker_id,
+        )
+        return task
+
+    def get_epoch(self):
+        return self._dataset_splitter.get_epoch()
+
+    def completed(self):
+        return (
+            self._dataset_splitter.epoch_finished()
+            and not self.todo
+            and not self.doing
+        )
+
+    def _create_todo_tasks(self, shards):
+        tasks = []
+        for shard in shards:
+            task = Task(self._task_id, self._task_type, shard)
+            tasks.append(task)
+            self._task_id += 1
+
+        logger.info(
+            "todo.extend: %d tasks created for dataset = %s.",
+            len(tasks),
+            self._dataset_splitter.dataset_name,
+        )
+        self.todo.extend(tasks)
+
+    def report_task_status(self, task_id, success):
+        doing_task = self.doing.pop(task_id)
+        if not doing_task:
+            logger.warning(
+                "Unknown task_id: %d of dataset %s"
+                % (task_id, self._dataset_splitter.dataset_name)
+            )
+            success = False
+        elif not success:
+            logger.warning(
+                "Task %d of %s failed "
+                % (task_id, self._dataset_splitter.dataset_name)
+            )
+            self.recover_task(doing_task.task)
+        else:
+            self._update_completed_step(doing_task.task)
+            logger.info(
+                "Task:%d completed, %d remaining tasks for Dataset %s",
+                task_id,
+                len(self.todo) + len(self.doing),
+                self._dataset_splitter.dataset_name,
+            )
+            task_completed_time = time.time() - doing_task.start_time
+            if task_completed_time > self._max_task_completed_time:
+                self._max_task_completed_time = task_completed_time
+        return success, doing_task
+
+    def _update_completed_step(self, task: Task):
+        record_count = task.shard.end - task.shard.start
+        batch_count = math.ceil(record_count / self._batch_size)
+        self._completed_step += batch_count
+
+    def get_completed_step(self):
+        return self._completed_step
+
+    def recover_task(self, task):
+        if not self._check_exceed_max_task_retries(task):
+            self.todo.append(task)
+
+    def _check_exceed_max_task_retries(self, task: Task):
+        task.retry_count += 1
+        if task.retry_count > _MAX_TASK_RETRIES:
+            logger.error(
+                "A task %s of failed with %d retries "
+                % (task.shard.name, _MAX_TASK_RETRIES)
+            )
+            return True
+        return False
+
+    def get_doing_tasks(self):
+        return self.doing
+
+    def get_task_from_todo(self, worker_id):
+        for shard in self.todo:
+            if shard.partition == worker_id:
```

Comment:
1. shard -> task because the todo stores Tasks. 
2. There is no `partition` in the Task class.